opentelemetry-api/opentelemetry/backend.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+The OpenTelemetry backend module defines and implements the API to access an
+implementation (backend) of the OpenTelemetry API.
+
+By default, if you call a getter function (e.g., :func:`tracer`) and the corresponding setter (e.g.,
+:func:`set_tracer`) wasn't called, you will get a default implementation, which is selected as follows:
+
+    1. If the environment variable OPENTELEMETRY_PYTHON_BACKEND_<getter-name>` (e.g.,
+       OPENTELEMETRY_PYTHON_BACKEND_TRACER) is set to an nonempty value, an attempt is made to
+       import a module with that name and call a function `get_opentelemetry_backend_impl` in it.
+       The function receives the API type that is expected as an argument and should return an
+       instance of it (e.g., the argument is :cls:`opentelemetry.trace.Tracer` and the function should
+       return an instance of a :cls:`~Tracer` (probably of a derived type).
+    2. If the variable is not set, `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` is tried instead.
+    3. If that variable was also not set, an attempt is made to import and use the OpenTelemetry
+       SDK.
+    4. Otherwise (if no variable was set and the SDK was not importable, or an error occured when
+       trying to instantiate the implementation object) the default implementation that ships with the
+        API distribution (a fast no-op implementation) is used.
+
+If you called the setter or if you call the setter later (not recommended), the object you set is
+used instead.
+"""
+
+import sys
+import os
+import importlib
+from typing import TypeVar, Type, Callable
+from functools import wraps
+
+from .trace import Tracer
+
+### Generic private code {{{1
+
+_T = TypeVar('_T')
+
+_DEFAULT_BACKEND_MODNAME = 'opentelemetry.sdk.internal.backend_impl'
+_UNIT_TEST_IGNORE_ENV = False
+
+_tracer = None
+
+def _get_fallback_impl(api_type: Type[_T]) -> _T:
+    # TODO: Move (most of) this to module docstring.
+    """Gets the fallback implementation for ``api_type``.
+
+    ``api_type`` must be a OpenTelemetry API type like :cls:`Tracer`.
+
+    First, the function tries to find a module that provides a `get_opentelemetry_backend_impl`
+    function (with the same signature as this function). The following modules are tried:
+
+    1. `$OPENTELEMETRY_PYTHON_BACKEND_<typename>` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
```

Comment:
Nit, but why is this module called "backend"? This could be confusing since "backend" usually refers to a vendor.

Reply:
This module *does* provide access to a vendor's implementation, at least that's the intent. The user would just use `set_tracer(MyVendorTracer(myoption='myvalue'))` or the vendor implementation is configured e.g. by a PaaS via the environment variable.

Reply:
I mean that it refers to the vendor's backend service. We say that the exporters send data to a given backend, as in over the wire to dynatrace, stackdriver, azure monitoring, etc.

Reply:
Ah, I get you now. Then what about the package name `implementation`? Or, if we split the loader and the global objects: `loader`.

Reply:
`loader` is my favorite, especially since `implementation` sounds like it could be the SDK package itself.

Reply:
It's now called `loader`.